lib/Target/TfheRust/TfheRustEmitter.cpp

```diff
@@ -221,14 +305,35 @@ LogicalResult TfheRustEmitter::printSksMethod(
     std::string_view op, SmallVector<std::string> operandTypes) {
   emitAssignPrefix(result);
 
-  auto operandTypesIt = operandTypes.begin();
+  if (!operandTypes.empty()) {
+    assert(operandTypes.size() == nonSksOperands.size() &&
+           "invalid sizes of operandTypes");
+    operandTypes =
+        llvm::to_vector(llvm::map_range(operandTypes, [&](std::string value) {
+          return value.empty() ? "" : " as " + value;
+        }));
+  }
+  auto *operandTypesIt = operandTypes.begin();
   os << variableNames->getNameForValue(sks) << "." << op << "(";
   os << commaSeparatedValues(nonSksOperands, [&](Value value) {
-    const auto *prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
-    return prefix + variableNames->getNameForValue(value) +
-           (!operandTypes.empty() ? " as " + *operandTypesIt++ : "");
+    auto valueStr = variableNames->getNameForValue(value);
+    if (isa<LookupTableType>(value.getType())) {
+      valueStr = "luts[\"" + variableNames->getNameForValue(value) + "\"]";
+    }
+    std::string prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
+    std::string suffix = operandTypes.empty() ? "" : *operandTypesIt++;
+    return prefix + valueStr + suffix;
   });
+
   os << ");\n";
+
+  // Insert ciphertext results into temp_nodes so that the levelled ops can
+  // reference them.
+  if (isa<EncryptedUInt3Type>(result.getType())) {
+    os << llvm::formatv("temp_nodes.insert({0}, {1}.clone());\n",
```

Comment:
I feel like this clone can be removed. If it's the only use of the output of an op, then it can use move semantics (or reference semantics) to go into the temp_nodes map, and no clone op is necessary. Though I may have to see the output program in more detail to confirm this...

Reply:
a clone can't be removed here because sometimes a value is used later - and then we have the value moved. i think it might be possible to inspect whether any of the uses are NOT an memref load which pull from temp_nodes and have it conditional on that - but for simplicity i'll keep the clone in and work on optimizing these little details after